spec/Section 3 -- Type System.md

```diff
@@ -1340,14 +1357,280 @@ Union type extensions have the potential to be invalid if incorrectly defined.
 
 1. The named type must already be defined and must be a Union type.
 2. The member types of a Union type extension must all be Object base types;
-   Scalar, Interface and Union types must not be member types of a Union.
-   Similarly, wrapping types must not be member types of a Union.
+   Scalar, Interface, Union and Tagged types must not be member types of a
+   Union.  Similarly, wrapping types must not be member types of a Union.
 3. All member types of a Union type extension must be unique.
 4. All member types of a Union type extension must not already be a member of
    the original Union type.
 5. Any non-repeatable directives provided must not already apply to the
    original Union type.
 
+
+## Tagged Types
+
+TaggedTypeDefinition : Description? tagged Name Directives[Const]? TaggedMembersDefinition?
+
+TaggedTypeExtension :
+  - extend tagged Name Directives[Const]? TaggedMembersDefinition
+  - extend tagged Name Directives[Const]
+
+TaggedMembersDefinition : { TaggedMemberDefinition+ }
+
+TaggedMemberDefinition : Description? Name : Type Directives[Const]?
+
+Tagged types represent a list of possible named members, exactly one of which
+must be present. This resulting member must yield a value of a specific type.
+In outputs, like for Objects, Tagged types should be serialized as ordered
+maps: the queried member names (or aliases) that are present are the keys and
+the result of evaluating the member is the value, ordered by the order in which
+they appear in the query. Note that Tagged types may be objects with more than
+one key even though they guarantee that exactly one member is present, this is
+because aliases may be used, and `__typename` may also be queried (and will
+return the name of the tagged type).
+
+All members defined within a Tagged type must not have a name which begins with
+{"__"} (two underscores), as this is used exclusively by GraphQL's
+introspection system.
+
+For example, a type `StringFilter` could be described as:
```

Comment:
There might be an example that makes the intent of allowing only one of the fields more immediately clear. Here, passing both `lengthAtLeast` and `lengthAtMost` seems reasonable.

How about something like this:

```graphql
tagged Angle {
  degree: Float!
  radians: Float!
}
```

Reply:
I definitely see your point. I'll try and think of a mutually-exclusive example which uses mixed scalars.

Reply:
Yeah, the mixed scalars illustrate the point better.

Reply:
I've added a type:

```graphql example
tagged PackQuantity {
  volumeInMetresCubed: Float!
  numberOfBoxes: Int!
  description: String!
}
```

I might be moving house soon :wink: